types/src/main.ts

```diff
@@ -2717,6 +2730,103 @@ class FullTextDocument implements TextDocument {
 	}
 }
 
+/**
+ * Incrementally-synchronized Text Document implementation that leverages a
+ * "Piece Tree" as text buffer. See the following for details:
+ *  - https://github.com/rebornix/PieceTree
+ *  - https://code.visualstudio.com/blogs/2018/03/23/text-buffer-reimplementation
+ *
+ * This text document implementation only supports "Incremental" text document
+ * sync. If you wish to use "Full" text document sync, use FullTextDocument, as it
+ * uses a simple JS string as an underlying text buffer, since the full text gets
+ * swapped out on every update.
+ */
+class IncrementalTextDocument implements TextDocument {
+
+	private _uri: string;
+	private _languageId: string;
+	private _version: number;
+	private _tree: PieceTreeBase;
+
+	public constructor(uri: string, languageId: string, version: number, content: string) {
+		this._uri = uri;
+		this._languageId = languageId;
+		this._version = version;
+
+		// Prepare Piece Tree
+		const ptBuilder = new PieceTreeTextBufferBuilder();
+		ptBuilder.acceptChunk(content);
```

Comment:
@rebornix - should I be slicing this up into 64 KB chunks, as you mentioned is done by `fs.readFile` in your [blog post](https://code.visualstudio.com/blogs/2018/03/23/text-buffer-reimplementation), or is this an appropriate use of your API (passing the entire `content` "chunk" in on instantiation).

Reply:
If the document comes as a single chunk of string, then we just use that. If it's super large, then the performance of editing at the very beginning is somewhat costly (as checking char code is slow on large strings (100+MB)) but it won't be worse than the original approach.